misc/dbt-materialize/dbt/include/materialize/macros/columns.sql

```diff
@@ -0,0 +1,24 @@
+-- Copyright Materialize, Inc. and contributors. All rights reserved.
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License in the LICENSE file at the
+-- root of this repository, or online at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+{% macro materialize__get_empty_subquery_sql(select_sql, select_sql_header=none) %}
+    -- The default macro generates a `SELECT * FROM(<query>) WHERE FALSE LIMIT
+    -- 0`, which in PostgreSQL is optimized to a no-op that returns no rows. In
+    -- Materialize, `SELECT ... LIMIT 0` requires a cluster. Because the
+    -- connection is not guaranteed to have a valid cluster, we pass the model
+    -- and header statements as raw SQL.
+    {% set sql = ({"select_sql": select_sql,"header_sql": select_sql_header}) %}
+    {{ return(sql) }}
```

Comment:
This looks to me like it will break the `get_columns_in_query` macro, which relies on this macro to return a single string, rather than a dict. Also I think this macro is technically public API, so downstream dbt projects/packages could be relying on the current behavior of this macro.

One alternative, though gross, could be to change the implementation of this macro to be:

```jinja2
{% macro materialize__get_empty_subquery_sql(select_sql, select_sql_header=none) %}
    {% do adapter.set_last_sql_header(select_sql_header) %}
    {{ default__get_empty_subquery_sql(select_sql) }}
{% endmacro %}
```

Then in `impl.py`:

```py
    def set_last_sql_header(select_sql_header: str):
        self.last_sql_headers[self.connections.get_thread_identifier()] = select_sql_header
```

And then, finally, you could pull the SQL header out of `self.last_sql_headers` in `get_column_schema_from_query`, even though the SQL header isn't passed in to the method directly.

Reply:
Good point about this being part of the public API (though 🫠). It's tough to not implement something gross without support for running both statements in a transaction.